Update API of `ServerlessHub<T>` according to the review meeting

Y-Sindo opened the pull request:
Review meeting link: https://github.com/Azure/azure-sdk/issues/3661
API VIEW link:  https://apiview.dev/Assemblies/Review/18274553e71a4f44a2f924031f627b14?diffRevisionId=2cfcb5bf94b0487daec57d5e35cb7b0e&doc=False&diffOnly=False
* Add `HubName` to keep consistent with untyped one. 
 * Make `GetClaims` method an instance method in case it needs to access instance members in the future.
 
## Newly added
* Add `SignalRTriggerEvents`,  `SignalRTriggerCategories`
* Update Azure SignalR SDK to 1.15.0

Y-Sindo committed:
Update API of ServerlessHub<T> according to the review meeting

azure-sdk commented:
API changes have been detected in `Microsoft.Azure.WebJobs.Extensions.SignalRService`. You can review API changes [here](https://apiview.dev/Assemblies/Review/a55b0c4728bb4765ae03b4b19a67a0a5?diffOnly=True&diffRevisionId=3cdec19b3b0f4ae1865f18ab2cc7c15a)

**API changes**
```diff
+         public string HubName { get; }
-         protected static IList<Claim> GetClaims(string jwt);
+         protected IList<Claim> GetClaims(string jwt);
```

KrzysztofCwalina commented on sdk/signalr/Microsoft.Azure.WebJobs.Extensions.SignalRService/api/Microsoft.Azure.WebJobs.Extensions.SignalRService.net5.0.cs:
```diff
@@ -0,0 +1,272 @@
+namespace Microsoft.Azure.WebJobs.Extensions.SignalRService
+{
+    public static partial class Category
```
Could we call it SignalRCategory and SignalREvent? Type names like Event and Category are too generic.

Y-Sindo replied:
They are already released classes. How about marking them as deprecated and adding new classes named SignalRCategory and SignalREvent?

Y-Sindo committed:
Rename `Event` -> `SignalRTriggerEvents`, `Category` -> `SignalRTriggerCategories`

azure-sdk commented:
API changes have been detected in `Microsoft.Azure.WebJobs.Extensions.SignalRService`. You can review API changes [here](https://apiview.dev/Assemblies/Review/a55b0c4728bb4765ae03b4b19a67a0a5?diffOnly=True&diffRevisionId=3cdec19b3b0f4ae1865f18ab2cc7c15a)

azure-sdk commented:
API changes have been detected in `Microsoft.Azure.WebJobs.Extensions.SignalRService`. You can review API changes [here](https://apiview.dev/Assemblies/Review/a55b0c4728bb4765ae03b4b19a67a0a5?diffOnly=True&diffRevisionId=3cdec19b3b0f4ae1865f18ab2cc7c15a)

Y-Sindo committed:
Update Azure SignalR SDK to 1.15.0

Why TFM .netcoreapp3.1 deleted: The `JsonHubProtocol` of ASP.NET SignalR uses `Newtonsoft.Json` for netcoreapp 2.1 and uses `System.Text.Json` for netcore3.1 and above. However, function host always uses `Newtonsoft.Json` to parse JSON data sent from isolated worker, and SignalR fails to serialize `Newtonsoft.Json.JObject` on .NET Core 3.1 correctly. In the past, SignalR SDK and extensions enabled `NewtonsoftJsonHubProtocol` (not available on .NET Core 2.1) on .NET Core 3.1 with isolated function workers only, so they had to use framework-sensitive conditional compilation. Since SignalR SDK 1.15.0, we have a new hub protocol accepting a `ObjectSerializer`, which is available on .NET Standard 2.0, so framework-sensitive conditional is unecessary.

JoshLove-msft commented on sdk/signalr/Microsoft.Azure.WebJobs.Extensions.SignalRService/Microsoft.Azure.WebJobs.Extensions.SignalRService.sln:
```diff
@@ -6,6 +6,7 @@ MinimumVisualStudioVersion = 15.0.26124.0
 Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Solution Items", "Solution Items", "{FFE2667E-131F-40CE-8851-AFAECFBF89BB}"
 	ProjectSection(SolutionItems) = preProject
 		Directory.Build.props = Directory.Build.props
+		..\..\..\eng\Packages.Data.props = ..\..\..\eng\Packages.Data.props
```
nit: remove

Y-Sindo replied:
I think it is more convenient to include the file in a solution so that we can update dependencies in a solution.

JoshLove-msft commented on sdk/signalr/Microsoft.Azure.WebJobs.Extensions.SignalRService/src/Category.cs:
```diff
@@ -1,12 +1,15 @@
     /// A class contains constant strings that represent SignalR trigger category.
     /// </summary>
     [System.Diagnostics.CodeAnalysis.SuppressMessage("Usage", "AZC0012:Avoid single word type names", Justification = "Breaking change")]
+    [Obsolete("Please use SignalRTriggerCategories instead.")]
```
We don't obsolete stuff due to the potential for it to cause breaking changes. Instead we use the EditorBrowsable attribute and set to Never.

JoshLove-msft commented on sdk/signalr/Microsoft.Azure.WebJobs.Extensions.SignalRService/src/Config/IConfigurationExtensions.cs:
```diff
@@ -65,5 +68,45 @@ public static bool TryGetNamedEndpointFromIdentity(this IConfigurationSection se
+
+        public static bool TryGetJsonObjectSerializer(this IConfiguration configuration, out ObjectSerializer serializer)
+        {
+            var hubProtocolConfig = configuration[Constants.AzureSignalRHubProtocol];
```
nit: check this later when it will be used?

JoshLove-msft commented on sdk/signalr/Microsoft.Azure.WebJobs.Extensions.SignalRService/api/Microsoft.Azure.WebJobs.Extensions.SignalRService.netstandard2.0.cs:
```diff
@@ -104,23 +106,23 @@ public abstract partial class ServerlessHub : System.IDisposable
         protected System.Collections.Generic.IList<System.Security.Claims.Claim> GetClaims(string jwt) { throw null; }
         protected Microsoft.Azure.WebJobs.Extensions.SignalRService.SignalRConnectionInfo Negotiate(string userId = null, System.Collections.Generic.IList<System.Security.Claims.Claim> claims = null, System.TimeSpan? lifetime = default(System.TimeSpan?)) { throw null; }
-        protected System.Threading.Tasks.Task<Microsoft.Azure.WebJobs.Extensions.SignalRService.SignalRConnectionInfo> NegotiateAsync(Microsoft.Azure.SignalR.Management.NegotiationOptions options) { throw null; }
+        protected virtual System.Threading.Tasks.Task<Microsoft.Azure.WebJobs.Extensions.SignalRService.SignalRConnectionInfo> NegotiateAsync(Microsoft.Azure.SignalR.Management.NegotiationOptions options) { throw null; }
```
This could be a subtle breaking change, was this agreed on at arch board?

Y-Sindo replied:
>❌ DISALLOWED: Adding the virtual keyword to a member
While this often is not a breaking change because the C# compiler tends to emit callvirt Intermediate Language (IL) instructions to call non-virtual methods (callvirt performs a null check, while a normal call doesn't), this behavior is not invariable for several reasons:
C# is not the only language that .NET targets.
The C# compiler increasingly tries to optimize callvirt to a normal call whenever the target method is non-virtual and is probably not null (such as a method accessed through the ?. null propagation operator).
Making a method virtual means that the consumer code would often end up calling it non-virtually.

https://docs.microsoft.com/en-us/dotnet/core/compatibility/





Seems the only consequence is that when we call the virtual method from the base class, it might work as non-virtual. I think that's OK because currently calling many methods of  `ServerlessHub<T>` from a `ServerlessHub` is always not working now and users will get null pointer exception as we pass a null `ServiceHubContext` to the base class. `ServiceHubContext` cannot be converted from `ServiceHubContext<T>` yet and we might improve it later.

JoshLove-msft commented on sdk/signalr/Microsoft.Azure.WebJobs.Extensions.SignalRService/api/Microsoft.Azure.WebJobs.Extensions.SignalRService.netstandard2.0.cs:
```diff
@@ -104,23 +106,23 @@ public abstract partial class ServerlessHub : System.IDisposable
-        protected System.Threading.Tasks.Task<Microsoft.Azure.WebJobs.Extensions.SignalRService.SignalRConnectionInfo> NegotiateAsync(Microsoft.Azure.SignalR.Management.NegotiationOptions options) { throw null; }
+        protected ServerlessHub() : base (default(Microsoft.Azure.SignalR.Management.IServiceHubContext), default(Microsoft.Azure.SignalR.Management.IServiceManager)) { }
+        protected ServerlessHub(Microsoft.Azure.SignalR.Management.ServiceHubContext<T> serviceHubContext) : base (default(Microsoft.Azure.SignalR.Management.IServiceHubContext), default(Microsoft.Azure.SignalR.Management.IServiceManager)) { }
+        public new Microsoft.Azure.SignalR.Management.ClientManager ClientManager { get { throw null; } }
```
Since we are making another method virtual, any reason to not make these properties virtual and override instead?

JoshLove-msft replied:
This comment applies to all of the properties where we are using `new`

Y-Sindo replied:
The properties other than `ClientManager` have different types between `ServerlessHub` and `ServerlessHub<T>`.  The base class always uses interface types and the derived class uses abstract classes. The `new` keyword seems necessary.

JoshLove-msft commented on sdk/signalr/Microsoft.Azure.WebJobs.Extensions.SignalRService/src/Event.cs:
```diff
@@ -1,12 +1,15 @@
     /// A class contains constant strings that represents different trigger event type.
     /// </summary>
     [System.Diagnostics.CodeAnalysis.SuppressMessage("Usage", "AZC0012:Avoid single word type names", Justification = "Breaking change")]
+    [Obsolete("Please use SignalRTriggerEvents instead.")]
```
Use EditorBrowsable never instead.

JoshLove-msft commented on sdk/signalr/Microsoft.Azure.WebJobs.Extensions.SignalRService/api/Microsoft.Azure.WebJobs.Extensions.SignalRService.netstandard2.0.cs:
```diff
@@ -1,10 +1,12 @@
         public const string Connections = "connections";
         public const string Messages = "messages";
     }
+    [System.ObsoleteAttribute("Please use SignalRTriggerEvents instead.")]
```
Use EditorBrowsable never instead.

JoshLove-msft commented on sdk/signalr/Microsoft.Azure.WebJobs.Extensions.SignalRService/api/Microsoft.Azure.WebJobs.Extensions.SignalRService.netstandard2.0.cs:
```diff
@@ -1,10 +1,12 @@
 namespace Microsoft.Azure.WebJobs.Extensions.SignalRService
 {
+    [System.ObsoleteAttribute("Please use SignalRTriggerCategories instead.")]
```
Use EditorBrowsable never instead.

Y-Sindo committed:
Update

azure-sdk commented:
API changes have been detected in `Microsoft.Azure.WebJobs.Extensions.SignalRService`. You can review API changes [here](https://apiview.dev/Assemblies/Review/a55b0c4728bb4765ae03b4b19a67a0a5?diffOnly=True&diffRevisionId=3cdec19b3b0f4ae1865f18ab2cc7c15a)

JoshLove-msft commented on sdk/signalr/Microsoft.Azure.WebJobs.Extensions.SignalRService/src/TriggerBindings/ServerlessHubOfT.cs:
```diff
@@ -1,58 +1,64 @@
+        /// Creates an instance of strongly typed serverless hub.
+        /// </summary>
+        /// <param name="serviceHubContext">Injects a service hub context.</param>
+        protected ServerlessHub(ServiceHubContext<T> serviceHubContext) : base(default(ServiceHubContext))
```
This seems less than ideal, since it means that if someone has a `ServerlessHub<T>` typed as a `ServerlessHub`, the context will not be correct. What would happen in that case when using any of the properties that we are hiding - Clients, Groups, and UserGroups?

JoshLove-msft replied:
Chatted offline, and the base type properties would actually throw nullref exceptions with the current implementation.

JoshLove-msft commented on sdk/signalr/Microsoft.Azure.WebJobs.Extensions.SignalRService/api/Microsoft.Azure.WebJobs.Extensions.SignalRService.netstandard2.0.cs:
```diff
@@ -104,23 +106,23 @@ public abstract partial class ServerlessHub : System.IDisposable
-        protected System.Threading.Tasks.Task<Microsoft.Azure.WebJobs.Extensions.SignalRService.SignalRConnectionInfo> NegotiateAsync(Microsoft.Azure.SignalR.Management.NegotiationOptions options) { throw null; }
+        protected ServerlessHub() : base (default(Microsoft.Azure.SignalR.Management.IServiceHubContext), default(Microsoft.Azure.SignalR.Management.IServiceManager)) { }
+        protected ServerlessHub(Microsoft.Azure.SignalR.Management.ServiceHubContext<T> serviceHubContext) : base (default(Microsoft.Azure.SignalR.Management.IServiceHubContext), default(Microsoft.Azure.SignalR.Management.IServiceManager)) { }
+        public override Microsoft.Azure.SignalR.Management.ClientManager ClientManager { get { throw null; } }
```
Does it make sense that this one overrides where the other properties hide? If someone has a `ServerlessHub<T>` typed as a `ServerlessHub` will the behavior be consistent between the different properties?

Y-Sindo committed:
Add `SignalRConnectionAttribute` for `ServerlessHub<T>`

Y-Sindo committed:
Fix for SignalRConnectionAttribute

JoshLove-msft commented on sdk/signalr/Microsoft.Azure.WebJobs.Extensions.SignalRService/src/TriggerBindings/SignalRConnectionAttribute.cs:
```diff
@@ -0,0 +1,30 @@
+        /// Initializes a new instance of the <see cref="SignalRConnectionAttribute"/> class.
+        /// </summary>
+        /// <param name="conneciton">Gets or sets the app setting name that contains the Azure SignalR connection.</param>
+        public SignalRConnectionAttribute(string conneciton)
```
```suggestion
        public SignalRConnectionAttribute(string connection)
```

JoshLove-msft commented:
I think you need to regenerate the API

JoshLove-msft commented on sdk/signalr/Microsoft.Azure.WebJobs.Extensions.SignalRService/src/TriggerBindings/ServerlessHubOfT.cs:
```diff
@@ -42,15 +43,29 @@ public abstract class ServerlessHub<T> where T : class
         /// Get the name of the hub.
         /// </summary>
-        public string HubName => typeof(T).Name;
+        public string HubName => GetType().Name;
```
Why did we make this change? Won't this give us `ServerlessHub<Foo>` rather than `Foo`? Is that what we want?

Y-Sindo replied:
This is by design. If users have a `ChatHub : ServerlessHub<IChatClient>`, then the hub name is `ChatHub` other than `IChatClient`. This is consistent with `ServerlessHub` and the the `Hub` in SignalR.

Y-Sindo committed:
Fix

azure-sdk commented:
API changes have been detected in `Microsoft.Azure.WebJobs.Extensions.SignalRService`. You can review API changes [here](https://apiview.dev/Assemblies/Review/a55b0c4728bb4765ae03b4b19a67a0a5?diffOnly=True&diffRevisionId=3cdec19b3b0f4ae1865f18ab2cc7c15a)

JoshLove-msft reviewed (approved):
Looks good!

Y-Sindo committed:
Remove unused using.
